include/revng/Pipeline/Container.h

```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <memory>
+
+#include "revng/Pipeline/Target.h"
+
+namespace Pipeline {
+
+class ContainerBase {
+protected:
+  char const *ID;
+
+public:
+  ContainerBase(char const *const ID) : ID(ID) {}
+  virtual std::unique_ptr<ContainerBase>
+  cloneFiltered(const TargetsList &Targets) const = 0;
+  virtual void mergeBack(ContainerBase &&) = 0;
+  virtual TargetsList enumerate() const = 0;
+  virtual bool remove(const TargetsList &Targets) = 0;
+  virtual llvm::Error storeToDisk(llvm::StringRef Path) const = 0;
+  virtual llvm::Error loadFromDisk(llvm::StringRef Path) = 0;
+  virtual ~ContainerBase() = default;
+
+  static bool classof(const ContainerBase *) { return true; }
+
+  template<typename Derived>
+  bool isA() const {
+    return ID == &Derived::ID;
+  }
+};
+
+/// crtp class to be extended to implement a pipeline container
+/// The methods that must be implemented are those shown in containerBase
+template<typename Derived>
+class Container : public ContainerBase {
+public:
+  Container() : ContainerBase(&Derived::ID) {}
+
+  void mergeBack(ContainerBase &&Container) final {
+    revng_assert(llvm::isa<Derived>(Container));
+    mergeBackDerived(std::move(llvm::cast<Derived>(Container)));
+  }
```

Comment:
crtp is not enough because the issue is that sometimes you have 2 containers you with to merge that you know have the same derived type, but don't know which type it is. Here is where one actually knows is own type and can cast the other guy as well.

Reply:
As an inheritor of this class, implementing mergeBack**Derived** makes me a bit suspicious.
Can we transform `mergeBackDerived` into an overload of `mergeBack`?
Otherwise, we could also drop `mergeBack` and rename `mergeBackDerived` into `mergeBack`. Then if users want to cast, they can do it.

Reply:
the issue with mergeBack instead of mergeBackDerived is that they would share the same name and accept as arguments the objects that are in a inheritance relationship, thus it would be really hard to notice if it starts to use the wrong overload due to implicit casts.

Reply:
Maybe we want to rename `mergeBackDerived` into `mergeBackImpl` and make it `protected`?